src/graphql/graphql-api.php

```diff
@@ -0,0 +1,245 @@
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockDataAttribute',
+			[
+				'description' => 'Block data attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => 'String',
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => 'String',
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual inner block, with the above attribute.
+		register_graphql_type(
+			'InnerBlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'         => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'parentId'   => [
+						'type'        => 'String',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. This will match the ID of the block',
+					],
+					'name'       => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes' => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'          => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'name'        => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes'  => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+					'innerBlocks' => [
+						'type'        => [ 'list_of' => 'InnerBlockData' ],
+						'description' => 'Flattened list of inner blocks of this block',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the list of individual blocks, with each item being the above type.
+		register_graphql_type(
+			'BlocksData',
+			[
+				'description' => 'Data for all the blocks',
+				'fields'      => [
+					'blocks'   => [
+						'type'        => [ 'list_of' => 'BlockData' ],
+						'description' => 'List of blocks data',
+					],
+					'warnings' => [
+						'type'        => [ 'list_of' => 'String' ],
+						'description' => 'List of warnings related to processing the blocks data',
+					],
+				],
+			],
+		);
+
+		// Register the field on every post type that supports 'editor'.
+		register_graphql_field(
+			'NodeWithContentEditor',
```

Comment:
I know you are following the pattern of the decoupled bundle here, but let's discuss the pros and cons of adding the field to this union.

Reply:
I had looked up [this](https://github.com/wp-graphql/wp-graphql/blob/develop/src/Type/InterfaceType/NodeWithContentEditor.php) page as well as [this](https://woographql.com/schema/definition/-node-with-content-editor) one to see what it could support. From what I gathered this would be what we support - pages and posts that would be made using the content editor. We have a check with the contentParser that already checks to see if the provided post has blocks within it or not, so that should account for the classic editor.

That was my thinking, alongside this being used in the decoupled bundle already. Was there something else or a downside that I might have missed in using this?

Reply:
Nevermind, I was conflating this with querying for `contentNodes`, which is a bit more problematic. I think this is fine.

> this would be what we support - pages and posts that would be made using the content editor.

The `NodeWithContentEditor` interface picks up [any post type that supports the content editor](https://github.com/wp-graphql/wp-graphql/blob/1051357687339e53b7bd19b816c756316dbc8a47/src/Registry/Utils/PostObject.php#L275-L277) regardless of its purpose or visibility. It's worth noting that the default value for `register_post_type#supports` includes `'editor'` (and also many developers and plugin authors don't pay close attention to post type args), so in reality the field will probably get added to a bunch of random plugin-created post types where it might not make sense. But it's fine since the user is unlikely to expose them in GraphQL or query for them. Just FYI.